compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala

```diff
@@ -65,6 +66,16 @@ class TreePickler(pickler: TastyPickler, attributes: Attributes) {
     fillRef(lengthAddr, currentAddr, relative = true)
   }
 
+  /** There are certain expectations with code which is naturally able to reach pickling
+   *  phase as opposed to one that uses best-effort compilation features.
+   *  When pickling betasty files, we do some custom checks, in case those expectations
+   *  cannot be fulfilled, and if then we can try to do something else.
+   *  For regular non best-effort compilation (without best-effort and without using .betasty on classpath),
+   *  this will always return true.
+   */
+  private inline def passesConditionForErroringBestEffortCode(condition: Boolean)(using Context): Boolean =
+    ((!ctx.isBestEffort && ctx.reporter.errorsReported) || ctx.usedBestEffortTasty) || condition
```

Comment:
Consider making the `condition` a by-name argument so that it does not get evaluated at all in normal runs:
```suggestion
  private inline def passesConditionForErroringBestEffortCode(condition: => Boolean)(using Context): Boolean =
    ((!ctx.isBestEffort && ctx.reporter.errorsReported) || ctx.usedBestEffortTasty) || condition
```

Reply:
Good point. Also while rechecking things I've noticed that I've put `!` here incorrectly. I remembered you might have signaled about that before, but I ended up just changing the name at that time. Now it's all correct